File: emm/testing_service/src/api/v1/endpoints/zephyr_status_mappings.py
"""`/zephyr-status-mappings` — маппинг статусов Zephyr → исход теста.

GET — действующий набор отдела (свой или по умолчанию), читает свой отдел.
PUT заменяет набор отдела целиком, DELETE возвращает к набору по умолчанию;
права — `(department_test_settings, *, update)` или department_admin отдела.
"""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from src.dependencies.auth import CurrentUserIdentity
from src.dependencies.db import get_db
from src.schemas.zephyr_status_mapping import ZephyrStatusMappingResponse, ZephyrStatusMappingUpdate
from src.services import zephyr_verdict as svc

router = APIRouter(prefix="/zephyr-status-mappings")


@router.get(
    "/{department_id}",
    response_model=ZephyrStatusMappingResponse,
    summary="Маппинг статусов Zephyr отдела",
    description=(
        "Какой статус тест-кейса в Zephyr считается пройденным, проваленным или "
        "незавершённым. `is_default=true` — у отдела своих строк нет."
    ),
    responses={403: {"description": "DEPARTMENT_ISOLATION — чужой отдел."}},
)
async def get_zephyr_status_mapping(
    department_id: str,
    identity: CurrentUserIdentity,
    db: AsyncSession = Depends(get_db),
) -> ZephyrStatusMappingResponse:
    return ZephyrStatusMappingResponse(**await svc.get_for(db, identity, department_id))


@router.put(
    "/{department_id}",
    response_model=ZephyrStatusMappingResponse,
    summary="Заменить маппинг статусов Zephyr отдела",
    description="Заменяет набор отдела целиком. Статусы уникальны без учёта регистра.",
    responses={403: {"description": "Нет права `department_test_settings:update`."}},
)
async def put_zephyr_status_mapping(
    department_id: str,
    body: ZephyrStatusMappingUpdate,
    identity: CurrentUserIdentity,
    db: AsyncSession = Depends(get_db),
) -> ZephyrStatusMappingResponse:
    return ZephyrStatusMappingResponse(**await svc.replace_for(db, identity, department_id, body))


@router.delete(
    "/{department_id}",
    response_model=ZephyrStatusMappingResponse,
    summary="Сбросить маппинг статусов Zephyr к набору по умолчанию",
    responses={403: {"description": "Нет права `department_test_settings:update`."}},
)
async def reset_zephyr_status_mapping(
    department_id: str,
    identity: CurrentUserIdentity,
    db: AsyncSession = Depends(get_db),
) -> ZephyrStatusMappingResponse:
    return ZephyrStatusMappingResponse(**await svc.reset_for(db, identity, department_id))

File: emm/web_ui/src/api/testing/zephyrStatusMappings.ts
/**
 * Тонкие обёртки над `testing_service` `/zephyr-status-mappings/*`.
 *
 * Какой статус тест-кейса в Zephyr считается пройденным, проваленным или
 * незавершённым (ждать дальше). У отдела без своих строк действует набор по
 * умолчанию (`is_default: true`). Чтение — свой отдел, запись — admin
 * testing_service или department_admin (`department_test_settings:update`).
 *
 * Source of truth: `testing_service/src/api/v1/endpoints/zephyr_status_mappings.py`.
 */

import { apiDelete, apiGet, apiPut } from "@/api/client";
import type { ZephyrStatusMapping, ZephyrStatusMappingItem } from "@/api/testing/types";

const BASE = "/testing/v1/zephyr-status-mappings";

/** `GET` — действующий маппинг отдела (свой или по умолчанию). */
export function getZephyrStatusMapping(departmentId: string): Promise<ZephyrStatusMapping> {
  return apiGet<ZephyrStatusMapping>(`${BASE}/${departmentId}`);
}

/** `PUT` — заменить набор отдела целиком. */
export function putZephyrStatusMapping(
  departmentId: string,
  items: ZephyrStatusMappingItem[],
): Promise<ZephyrStatusMapping> {
  return apiPut<ZephyrStatusMapping>(`${BASE}/${departmentId}`, { items });
}

/** `DELETE` — удалить строки отдела, вернуться к набору по умолчанию. */
export function resetZephyrStatusMapping(departmentId: string): Promise<ZephyrStatusMapping> {
  return apiDelete<ZephyrStatusMapping>(`${BASE}/${departmentId}`);
}
